View Listener

davedevelopment opened the pull request:
Previously #1062 and #863

TODO

- [x] Documentation

stof commented on src/Silex/Application.php:
```diff
@@ -89,7 +89,7 @@ public function __construct(array $values = array())
         $this['dispatcher_class'] = 'Symfony\\Component\\EventDispatcher\\EventDispatcher';
         $this['dispatcher'] = $this->share(function () use ($app) {
-            /**
+            /*
```
this should be kept as a docComment, as it is really one. the real fix would be to make it valid: ``@var EventDispatcherInterface $dispatcher``

davedevelopment replied:
Cheers @stof, learnt something new :)

stof commented on src/Silex/ViewListenerWrapper.php:
```diff
@@ -0,0 +1,88 @@
+     * Constructor.
+     *
+     * @param Application $app      An Application instance
+     * @param callable    $callback
```
this may not be a callable yet here (it is a callable only after being resolved)

stof commented on src/Silex/Application.php:
```diff
@@ -405,6 +405,23 @@ public function error($callback, $priority = -8)
+     * an instance of Response. When this occurs, all suitable handlers will be
+     * called, until one returns a Response object.
+     *
+     * @param callable $callback View handler callback
```
this is not (always) a callable here

davedevelopment replied:
Thanks @stof, fixed.

stof commented on tests/Silex/Tests/ApplicationTest.php:
```diff
@@ -559,6 +559,128 @@ public function testRedirectDoesNotRaisePHPNoticesWhenMonologIsRegistered()
+
+    public function testViewListenerWithCallableTypeHint()
+    {
+        if (version_compare(phpversion(), '5.4.0', '<')) {
```
use ``PHP_VERSION_ID`` here to compare the version (or ``@requires PHP 5.4`` in the docComment of the test, using the builtin PHPUnit feature)

davedevelopment commented:
Also worth mentioning @fabpot [disliked the typehinting aspect](https://github.com/silexphp/Silex/pull/1062#issuecomment-70408305). Happy to take it out, just need a final confirmation.

fabpot commented:
I'm fine with the typehinting if people like it.

Crell commented:
The type hinting is quite useful, IMO. It makes it very easy to return multiple object types from controllers and have the right corresponding view listener "just work".

The branch looks good to me visually, although IMO it could stand some more comments, particularly in shouldRun(). :smile:  The fabbot disagrees though.

henrikbjorn commented:
:-1: on the typehinting magick. Using a guard isnt that difficult to understand or explain. The magic typehinting is hard to explain and understand.

anlutro commented:
I like typehinting magic, but Silex doesn't really do it much elsewhere, so for the sake of consistency I'd be inclined to leave it out.

mediamonks-robert commented:
I +1 the typehinting since it's basically the same as the error handler and it's not hard to understand at all.

Crell commented:
As Robert said, it's the same logic and largely the same code structure as exception handling already uses.  If it's not confusing for exceptions it shouldn't be confusing for view(), either.

davedevelopment commented:
I don't really mind about having the type hinting per se, but I do think in it's present state, this PR is consistent with the existing error handler behaviour, which is itself a kind of special case of view handling. 

I've added some docs, comments on those appreciated as well.

fabpot commented on doc/usage.rst:
```diff
@@ -539,6 +539,58 @@ early::
+callable) to the view method. The callable should accept some sort of result
+from the controller::
+
+    $app->view(function (Response $response) {
```
I think this is unfortunate to use a Response here as ann example as this is probably not the common use case.

GromNaN replied:
A better example would be: 
```php
$app->view(function (array $controllerResult) use ($app) {
    return $app->json($controllerResult);
});
```

fabpot commented on doc/usage.rst:
```diff
@@ -539,6 +539,58 @@ early::
+        return $response;
+    });
+
+View Handlers also receive the ``Request`` as its second argument,
```
as their?

fabpot commented on doc/usage.rst:
```diff
@@ -539,6 +539,58 @@ early::
+-------------
+
+View Handlers allow you to intercept a controller result and transform it before
+it gets returned to the kernel.
```
You should mention that this is only when the controller result is not a response.

davedevelopment commented:
Replaced the ridiculously incorrect example, fixed the grammar and totally removed the last example, as I couldn't think of anything that wasn't a foobar type thing. Thanks @fabpot @GromNaN

fabpot commented:
Thank you @davedevelopment.
